src/CADL.Extension/Emitter.Csharp/src/lib/model.ts

```diff
@@ -407,27 +349,121 @@ export function getInputType(
         } as InputListType;
     }
 
-    function getInputTypeForMap(
-        program: Program,
-        key: Type,
-        value: Type
-    ): InputType {
+    function getInputTypeForMap(key: Type, value: Type): InputDictionaryType {
         return {
             Name: "Dictionary",
             KeyType: getInputType(program, key, models, enums),
             ValueType: getInputType(program, value, models, enums),
             IsNullable: false
         } as InputDictionaryType;
+    }   
+
+    function getInputModelForModel(m: ModelType): InputModelType {
+        m = getEffectiveSchemaType(program, m) as ModelType;
+        const name = getFriendlyName(program, m) ?? m.name;
+        let model = models.get(name);
+        if (!model) {
+            const baseModel = getInputModelBaseType(m.baseModel);
+            const properties: InputModelProperty[] = [];
+
+            model = {
+                Name: name,
+                Namespace: m.namespace?.name,
+                Description: getDoc(program, m),
+                IsNullable: false,
+                DiscriminatorPropertyName: getDiscriminator(program, m)?.propertyName,
+                DiscriminatorValue: getDiscriminatorValue(m, baseModel),
+                BaseModel: baseModel,
+                Properties: properties // Properties should be the last assigned to model
+            } as InputModelType;
+
+            models.set(name, model);
+
+            // Resolve properties after model is added to the map to resolve possible circular dependencies
+            addModelProperties(m.properties, properties, baseModel?.DiscriminatorPropertyName);
+
+            // Temporary part. Derived types may not be referenced directly by any operation
+            // We should be able to remove it when https://github.com/Azure/cadl-azure/issues/1733 is closed 
+            if (model.DiscriminatorPropertyName && m.derivedModels) {
+                m.derivedModels.forEach(dm => {
+                    getInputType(program, dm, models, enums);
+                });
+            }
+        }
+
+        return model;
+    }   
+
+    function getDiscriminatorValue(m: ModelType, baseModel?: InputModelType) : string | undefined {
+        const discriminatorPropertyName = baseModel?.DiscriminatorPropertyName;
+
+        if (discriminatorPropertyName) {
+            const discriminatorProperty = m.properties.get(discriminatorPropertyName);
+            if (discriminatorProperty?.type.kind === "String") {
+                return discriminatorProperty.type.value;
+            }
+        }
+
+        return undefined;
     }
-}
+
+    function addModelProperties(
+        inputProperties: Map<string, ModelTypeProperty>,
+        outputProperties: InputModelProperty[],
+        discriminatorPropertyName?: string
+    ) : void {
+        inputProperties.forEach(
+            (value: ModelTypeProperty, key: string) => {
+                if (value.name !== discriminatorPropertyName) {
```

Comment:
discriminator proeprty will not be added into the model. When use this function to resolve properties for base model, it seems that the discriminator property will be missing.

Reply:
Yes, that is intentional. Discriminator property is not really a property (essentially, it is an internal constant), existing code in HLC has to deal with it, not always successfully. I want to try to avoid it by not making it a property at all.

Reply:
Discriminator property will be an class Field of the base model. When we generate the model, codegen will add discriminator as a field together with other properties of the base model. OK. thanks @AlexanderSher

Reply:
Hello @AlexanderSher The discriminator property is missing in base model (see the generated test project). And in the test project, the discriminator even does not defined as a property in the base model in cadl file. Does cadl support the polymorphism that way which is different with API spec definition?

Reply:
We don't have finalized design for polymorphic models yet. https://github.com/Azure/autorest.csharp/issues/2361 still has `needs-design` tag:
![image](https://user-images.githubusercontent.com/8933858/186053144-49a17b76-6c05-4371-8cf3-f5d94439cb90.png)